scripts/shellcheck.sh

```diff
@@ -0,0 +1,253 @@
+                # valid if -S has any of 'info', 'warning', 'error'
+                elif [[ "${optarg}" == "info" || "${optarg}" == "warning" || "${optarg}" == "error" ]]; then
+                    SEVERITY="${optarg}"
+                else
+                    # continue and disregard invalid severity level
+                    printf "${RED}[Error]${RST_COLOR} invalid severity level: '${CYAN}%s${RST_COLOR}'\n" "${optarg}"
+                    _err=1
+                fi
+                shift "$shft_cnt"
+                ;;
+            -e*|--exclude-code*)
+                # strip whitespace from optarg
+                CODE=$(echo "${optarg}" | tr -d '[:space:]')
+                # valid if matching format SC1000-SC9999
+                if [[ "${optarg}" =~ ^SC[0-9]{4}$ ]]; then
+                    # if empty then populate with just error code, otherwise add pipe before new code for grep later
+                    if [[ -z "${EXCLUDE_CODES}" ]]; then
+                        EXCLUDE_CODES+="${CODE}"
+                    else
+                        EXCLUDE_CODES+=",${CODE}"
+                    fi
+                else
+                    # continue just don't save invalid error code
+                    printf "${RED}[Error]${RST_COLOR} Invalid error code entered: %s\n" "${optarg}"
+                    _err=1
+                fi
+                shift "$shft_cnt"
+                ;;
+            -v|--view)
+                VIEW="True"
+                shift "$shft_cnt"
+                ;;
+            -h|--help)
+                _help=1
+                ;;
+            *)
+                printf "${RED}[Error]${RST_COLOR} Unrecognized option: %s\n" "$1"
+                _err=1
+                shift "$shft_cnt"
+                ;;
+        esac
+
+        _invalid_arg_count=$(( "${_invalid_arg_count}" + 1 ))
+
+        # exit on help message
+        if [[ $_help -eq 1 ]]; then
+            show_usage
+            exit 0
+        fi
+        # continue on invalid arg, let user know but don't exit
+        if [[ "$_err" -eq 1 ]]; then
+            printf "${RED}[Error]${RST_COLOR} Invalid argument: %s\n" "${optarg}"
+        fi
+
+    done
+
+    if [[ "$_invalid_arg_count" -gt 0 ]]; then
+        # print the help message just once if one or more invalid arguments
+        echo; echo -e "${RED}[Error]${RST_COLOR} Some invalid arguments were detected."; echo
+        show_usage
+    fi
+}
+
+check_shellcheck_install() {
+    if ! command -v shellcheck &>/dev/null; then
+        echo -e "${RED}[ERROR]${RST_COLOR} shellcheck is not installed."
+        echo "Run 'sudo ./scripts/install-build-tools.sh' to install shellcheck."
+        exit 1
+    fi
+}
+
+get_num_cores() {
+    NUM_CORES=1
+    if [[ "$OSTYPE" == "linux-gnu"* ]]; then
+        NUM_CORES=$(grep -c ^processor /proc/cpuinfo)
+    elif [[ "$OSTYPE" == "darwin"* ]]; then
+        NUM_CORES=$(sysctl -n hw.ncpu)
+    else
+        NUM_CORES=1
+    fi
+}
+
+run_shellcheck() {
```

Comment:
`shellcheck` supports it's own severity filter (`-S`, the same as this script uses). Why do we not just forward that argument directly? A whole lot of this script could be simplified.

Reply:
ACK - did that as well, but added logic to count waived severities and error codes and summarize all